artichoke-backend/src/extn/core/time/subsec.rs

```diff
@@ -0,0 +1,562 @@
+            (b"0.0009".as_slice(), (0, 0)),
+        ];
+
+        let subsec_unit: Option<&[u8]> = Some(b":usec");
+
+        for (input, expectation) in &expectations {
+            let result = subsec(&mut interp, (Some(input), subsec_unit)).unwrap();
+            assert_eq!(
+                result.to_tuple(),
+                *expectation,
+                "Expected TryConvertMut<(Some({}), None), Result<Subsec>>, to return {} secs, {} nanos",
+                input.as_bstr(),
+                expectation.0,
+                expectation.1
+            );
+        }
+    }
+
+    #[test]
+    fn float_subsec_nanos() {
+        let mut interp = interpreter();
+
+        let expectations = [
+            // Numbers in and around 0.
+            (b"-1000000000.5".as_slice(), (-2, 999_999_999)),
+            (b"-1000000000.0".as_slice(), (-2, 0)),
+            (b"-999999999.5".as_slice(), (-1, 0)),
+            (b"-999999999.0".as_slice(), (-1, 1)),
+            (b"-1000.5".as_slice(), (-1, 999_998_999)),
+            (b"-1.5".as_slice(), (-1, 999_999_998)),
+            (b"-1.0".as_slice(), (-1, 999_999_999)),
+            (b"-0.0".as_slice(), (0, 0)),
+            (b"0.0".as_slice(), (0, 0)),
+            (b"1.0".as_slice(), (0, 1)),
+            (b"1.5".as_slice(), (0, 1)),
+            (b"1000.5".as_slice(), (0, 1_000)),
+            (b"999999999.0".as_slice(), (0, 999_999_999)),
+            (b"999999999.5".as_slice(), (0, 999_999_999)),
+            (b"1000000000.0".as_slice(), (1, 0)),
+            (b"1000000000.5".as_slice(), (1, 0)),
+            (b"1000000001.0".as_slice(), (1, 1)),
+            // Nanosecond and below (truncates, does not round).
+            (b"-0.1".as_slice(), (-1, 999_999_999)),
+            (b"0.1".as_slice(), (0, 0)),
+        ];
+
+        let subsec_unit: Option<&[u8]> = Some(b":nsec");
+
+        for (input, expectation) in &expectations {
+            let result = subsec(&mut interp, (Some(input), subsec_unit)).unwrap();
+            assert_eq!(
+                result.to_tuple(),
+                *expectation,
+                "Expected TryConvertMut<(Some({}), None), Result<Subsec>>, to return {} secs, {} nanos",
+                input.as_bstr(),
+                expectation.0,
+                expectation.1
+            );
+        }
+    }
+
+    #[test]
+    fn float_nan_raises() {
+        let mut interp = interpreter();
+
+        let err = subsec(&mut interp, (Some(b"Float::NAN"), None)).unwrap_err();
+
+        assert_eq!(err.name(), "FloatDomainError");
+        assert_eq!(err.message(), b"NaN".as_slice());
+    }
+
+    #[test]
+    fn float_infinite_raises() {
+        let mut interp = interpreter();
+
+        let err = subsec(&mut interp, (Some(b"Float::INFINITY"), None)).unwrap_err();
+
+        assert_eq!(err.name(), "FloatDomainError");
+        assert_eq!(err.message(), b"Infinity".as_slice());
+    }
+
+    #[test]
+    fn invalid_subsec_unit() {
```

Comment:
needs test for non-`Symbol` unit like `"bad_unit"` and `Object.new`.

Reply:
Filed:

- https://github.com/artichoke/artichoke/issues/2095#issuecomment-1221610741